lib/site.js

```diff
@@ -58,34 +63,43 @@ module.exports = class Site {
 			await mkdirp(this.target);
 		}
 
-		let { tree, layout } = this;
+		let { tree, renderers, generateLayout } = this;
 		if(!tree || filepaths === null) {
 			tree = await this.load();
 		}
 
+		let layout = generateLayout(tree);
+		let pageRenderer = new PageRenderer({ renderers, layout });
+
 		let dirs = new Set();
-		let writes = [];
-		tree.forEach(async page => {
-			let filepath = page.targetPath(this.target);
+		let writes = tree.map(async page => {
+			let filepath = path.resolve(this.target, page.slug);
 			if(filepaths && !filepaths.has(filepath)) {
 				return;
 			}
 
-			let dirPath = path.dirname(filepath);
-			if(!dirs.has(dirPath)) {
-				await mkdirp(dirPath);
-				dirs.add(dirPath);
+			if(!dirs.has(filepath)) {
+				await mkdirp(filepath);
+				dirs.add(filepath);
 			}
 
-			let op = page.render(layout).
-				then(html => {
-					// NB: intentionally avoiding faucet's asset manager because
-					//     we do not want fingerprinting or manifest handling
-					writeFile(filepath, html, "utf8");
-				});
-			writes.push(op);
+			page.filepath = filepath;
+
+			let html = await pageRenderer.render(page);
+
+			// NB: intentionally avoiding faucet's asset manager because
+			//     we do not want fingerprinting or manifest handling
+			// TODO: This also results in no output on the CLI incl. no status
```

Comment:
This sucks and we need a solution for this

Reply:
no comprendo 🤷‍♂️

Reply:
Whenever faucet-pipeline generates a file, it will create an output on the command line. If it was successful, it will give a checkmark, if it wasn't succesful it will show an error. If one of the files is an error file and you run without watch, it will also lead to the according exit status – while on watch it will keep on going and generate an error file.
All of this is gone if you just `writeFile` here.

Reply:
Oh, I see. Not sure the faucet semantics map to this task though, because a single source file generates many target files. 🤔 I guess you could argue it generates a single directory, but I'm not sure that makes this any easier.

Reply:
We could
a) elevate `AssetManager#_report` to a reusable utility function, much like we did with `resolvePath` _and_
b) reuse `AssetManager#exitOnError` here to `throw` if necessary.

Yes, that's a (tiny) bit of duplication, but you might also call it flexibility. 🙂

Reply:
Good idea 👍 Implemented in 20c6e7278066bcfe4b1afe23bc32846706274e28
What do you think? Obviously the "private" method should be made public to make this more obvious.

Reply:
Yeah, I was thinking `report(filepath, error)` in faucet-core's `util/index.js`. (Yes, "report" is a bit generic - blame [Karlton](https://www.martinfowler.com/bliki/TwoHardThings.html). ¯\\\_(ツ)_/¯ )

Reply:
I'm confused. I thought we just rename _report to report?

Reply:
Because OO ~is the devil~ should only be employed as a last resort - cf. https://github.com/faucet-pipeline/faucet-pipeline-core/pull/47.

Reply:
Oooh, ok. Now I get it. Let's do this after this PR, this is already gotten out of hand. Noted above as a follow up.